explicit-bzero/explicit-bzero.c

```diff
@@ -0,0 +1,106 @@
+/*
+ * Portable Snippets - https://github.com/nemequ/portable-snippets
+ * Created by arnavyc <arnavyc@outlook.com>
+ *
+ *   To the extent possible under law, the authors have waived all
+ *   copyright and related or neighboring rights to this code.  For
+ *   details, see the Creative Commons Zero 1.0 Universal license at
+ *   https://creativecommons.org/publicdomain/zero/1.0/
+ *
+ */
+
+#ifndef __STDC_WANT_LIB_EXT1__
+/* memset_s() function in C11 Annex K (bound-checking interface) */
+#define __STDC_WANT_LIB_EXT1__ 1
+#endif
+
+#include <stdlib.h>
+#include <string.h>
+
+#include "explicit-bzero.h"
+
+#if defined(__FreeBSD__) || defined(__OpenBSD__)
+#include <sys/param.h>
+#endif
+
+#ifdef _WIN32
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#endif /* _WIN32 */
+
+#ifdef __STDC_LIB_EXT1__
+/* memset_s() function in C11 Annex K (bound-checking interface) */
+#define HAVE_MEMSET_S 1
+#endif /* __STDC_LIB_EXT1__ */
+
+#if (defined(__GLIBC__) && __GLIBC_PREREQ(2, 25)) ||                           \
```

Comment:
The preprocessor doesn't short-circuit when evaluating conditions, so in this case the preprocessor will actually try to expand `__GLIBC_PREREQ(2, 25)` even if it doesn't exist, leading to an error on non-glibc platforms (I see this problem a lot with people trying to do stuff like `defined(__has_builtin) && __has_builtin(__builtin_memset)`).  You could do something like:

```c
#if defined(__GLIBC__)
  #if __GLIBC_PREREQ(2, 25)
    #define HAVE_EXPLICIT_BZERO
  #endif
#endif
#if !defined(HAVE_EXPLICIT_BZERO) && ( \
      (defined(__FreeBSD__) && __FreeBSD_version >= 1100037) || \
      (defined(__OpenBSD__) && OpenBSD >= 201405) \
    )
  #define HAVE_EXPLICIT_MEMSET
#endif
```

But that's pretty ugly.  I tend to prefer avoiding the macro altogether and just checking `__GLIBC__` and `__GLIBC_MINOR__`… the standard specifies that undefined object-like macros are replaced with 0, so it should be safe:

```c
#if \
    (defined(__GLIBC__) && (__GLIBC__ > 2 || (__GLIBC__ > 2 && __GLIBC_MINOR__ >= 25))) || \
    (defined(__FreeBSD__) && __FreeBSD_version >= 1100037) || \
    (defined(__OpenBSD__) && OpenBSD >= 201405) \
```